Haptics/app/src/main/java/com/example/android/haptics/samples/ui/views/HomeViewModel.kt

```diff
@@ -0,0 +1,257 @@
+                                HapticFeedbackButton(
+                                    application.getString(R.string.home_screen_slow_rise),
+                                    isPrimitiveSupported(vibrator, VibrationEffect.Composition.PRIMITIVE_SLOW_RISE),
+                                    VibrationEffect.Composition.PRIMITIVE_SLOW_RISE
+                                ),
+                                HapticFeedbackButton(
+                                    application.getString(R.string.home_screen_quick_rise),
+                                    isPrimitiveSupported(vibrator, VibrationEffect.Composition.PRIMITIVE_QUICK_RISE),
+                                    VibrationEffect.Composition.PRIMITIVE_QUICK_RISE
+                                ),
+                                HapticFeedbackButton(
+                                    application.getString(R.string.home_screen_quick_fall),
+                                    isPrimitiveSupported(vibrator, VibrationEffect.Composition.PRIMITIVE_QUICK_FALL),
+                                    VibrationEffect.Composition.PRIMITIVE_QUICK_FALL
+                                ),
+                                HapticFeedbackButton(
+                                    application.getString(R.string.home_screen_spin),
+                                    isPrimitiveSupported(vibrator, VibrationEffect.Composition.PRIMITIVE_SPIN),
+                                    VibrationEffect.Composition.PRIMITIVE_SPIN
+                                ),
+                                HapticFeedbackButton(
+                                    application.getString(R.string.home_screen_thud),
+                                    isPrimitiveSupported(vibrator, VibrationEffect.Composition.PRIMITIVE_THUD),
+                                    VibrationEffect.Composition.PRIMITIVE_THUD
+                                ),
+
+                            )
+                        ),
+                    )
+                )
+                return HomeViewModel(
+                    vibrator = vibrator, viewModelState = viewModelState,
+                    scaffoldState = scaffoldState, scrollState = scrollState
+                ) as T
+            }
+        }
+    }
+}
+
+/**
+ * Query whether the vibrator natively supports the a given effects.
+ * Note that this logic means that either the effect is definitely supported by the hardware,
+ * or it's status is unknown. Even with a value of VIBRATION_EFFECT_SUPPORT_NO, the device would
+ * either fall back to an unoptimized version or no vibration at all.
+ *
+ * See https://developer.android.com/reference/android/os/Vibrator#areEffectsSupported(int...)>
+ * for more information.
+ */
+private fun isEffectSupported(vibrator: Vibrator, effectId: Int): Boolean {
+    if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.R) return false
+    return vibrator.areAllEffectsSupported(effectId) !== Vibrator.VIBRATION_EFFECT_SUPPORT_NO
```

Comment:
This is a bit tricky, but effects are always going to play something because the platform have a fallback. Sometimes we can know that effects did play as defined by the device hardware (if they implement this method and return VIBRATION_EFFECT_SUPPORT_YES) but in most cases we can't know which effect played, if it was the device one or the fallback. I'd skip this check for effects and keep them all enabled.

Reply:
Removed and just check if isAndroidQOrLater